### PR TITLE
Blog App: Integration Specs

### PR DESCRIPTION
**Pull request by rmiyoyo**
# Work Done

#### n+1
- The N+1 problem is solved when fetching all posts and their comments for a user.
ca1
dfe

#### Integration specs
- Used Capybara to write integration tests for each view in the project.
- User index page:
  - I can see the username of all other users.
  - I can see the profile picture for each user.
  - I can see the number of posts each user has written.
  - When I click on a user, I am redirected to that user's show page.

- user show page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see the user's bio.
  - I can see the user's first 3 posts.
  - I can see a button that lets me view all of a user's posts.
  - When I click a user's post, it redirects me to that post's show page.
  - When I click to see all posts, it redirects me to the user's post's index page.

- User post index page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see a post's title.
  - I can see some of the post's body.
  - I can see the first comments on a post.
  - I can see how many comments a post has.
  - I can see how many likes a post has.
  - I can see a section for pagination if there are more posts than fit on the view.
  - When I click on a post, it redirects me to that post's show page.

- Post show page:
  - I can see the post's title.
  - I can see who wrote the post.
  - I can see how many comments it has.
  - I can see how many likes it has.
  - I can see the post body.
  - I can see the username of each commentor.
  - I can see the comment each commentor left.

### CODE REVIEW

File: spec/views/user_post_show_spec.rb [hunk 1]
```diff
@@ -3,8 +3,8 @@
 RSpec.describe 'Post', type: :system do
   before do
     @user1 = User.create(name: 'John Doe', photo: 'https://placehold.co/200x200',
```
Comment (thecodechaser) [inline spec/views/user_post_show_spec.rb]: - [ ] You missed adding the below test for the user show page, Please make sure to add that as requested in the project requirements.
  - Tests that assert you can see the user's first 3 posts.

Comment (thecodechaser) [inline spec/views/user_post_show_spec.rb]: - [ ] You missed adding the below tests for the post show page, Please make sure to add them as requested in the project requirements.
  - Tests that assert you can see who wrote the post.
  - Tests that assert you can see the username of each commenter.
  - Tests that assert you can see the comment each commentor left.

Comment (thecodechaser) [inline spec/views/user_post_show_spec.rb]: - [ ] All of the tests are failing, Please make sure that all tests pass without any errors.

image

**Review by thecodechaser — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ 

## Hi @rmiyoyo, and Team👋, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## To Highlight 🎉
- Good job fixing n+1 problem. ✔️ 
- Gitflow is followed. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- No linter errors. :heavy_check_mark:
- PR has a good title and summary. ✔️



## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by rmiyoyo** [comment]
# Work Done After Review
- All tests work
- All instances tested

File: spec/views/users_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  end
+
+  it 'displays post information on the User show page' do
+    visit user_path(id: @user1.id)
+    expect(page).to have_content(@post1.title)
+    expect(page).to have_content(@post1.text)
+    expect(page).to have_content(@post2.title)
+    expect(page).to have_content(@post2.text)
+    expect(page).to have_content('See all posts')
+  end
```
Comment (Ibizugbe) [inline spec/views/users_show_spec.rb]: Just [as suggested by the previous reviewer](https://github.com/rmiyoyo/blog-app/pull/11#discussion_r1455461342)😉, Kindly add a test that asserts that you can see the user's first 3 posts in the `user_show` page as this is a requirement for this milestone😉.

File: spec/views/user_post_show_spec.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  it 'displays post information on the User post show page' do
+    visit user_post_path(user_id: @user1.id, id: @post1.id)
+    expect(page).to have_content(@user1.name)
+    expect(page).to have_content(@post1.title)
+    expect(page).to have_content(@post1.text)
+    expect(page).to have_content(@user2.name)
+    expect(page).to have_content(@comment.text)
+    expect(page).to have_content('Likes: 0')
+    expect(page).to have_content('Comments: 1')
+  end
```
Comment (Ibizugbe) [inline spec/views/user_post_show_spec.rb]: [OPTIONAL] Nice job implementing the suggestions from the previous reviewer👍🏼, but I think we can break this `it` block to several `it` blocks to make your tests more descriptive and easier to understand😉.

**Review by Ibizugbe — CHANGES_REQUESTED** [review]
Hi @Team👋🏼👋🏼👋🏼, 

request change


Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights✅✅✅

- [x] Tests that assert you can see the author of the post, username of each commenter and their comments have been added👍🏼.


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by rmiyoyo** [comment]
# Work Done
I added the following tests:
1. Tests that assert you can see the user's first 3 posts.
2. Tests that assert you can see who wrote the post.
3. Tests that assert you can see the username of each commenter.
4. Tests that assert you can see the comment each commentor left.

**Review by fobadara — APPROVED** [review]
Hi team, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

APPRVED
## Highlights
✅ Descriptive README
✅ Descriptive PR summary
✅ Clean code

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**